io.openems.edge.ess.sma/src/io/openems/edge/ess/sma/stpxx3se/batteryinverter/BatteryInverterSmaStpSeImpl.java

```diff
@@ -0,0 +1,264 @@
+	private final CalculateEnergyFromPower calculateActiveDischargeEnergy = new CalculateEnergyFromPower(this,
+			SymmetricBatteryInverter.ChannelId.ACTIVE_DISCHARGE_ENERGY);
+	private final CalculateEnergyFromPower calculateDcChargeEnergy = new CalculateEnergyFromPower(this,
+			HybridManagedSymmetricBatteryInverter.ChannelId.DC_CHARGE_ENERGY);
+	private final CalculateEnergyFromPower calculateDcDischargeEnergy = new CalculateEnergyFromPower(this,
+			HybridManagedSymmetricBatteryInverter.ChannelId.DC_DISCHARGE_ENERGY);
+
+	private Config config;
+
+	public BatteryInverterSmaStpSeImpl() throws OpenemsException {
+		super(//
+				ACTIVE_MODELS, //
+				OpenemsComponent.ChannelId.values(), //
+				ModbusComponent.ChannelId.values(), //
+				SymmetricBatteryInverter.ChannelId.values(), //
+				ManagedSymmetricBatteryInverter.ChannelId.values(), //
+				HybridManagedSymmetricBatteryInverter.ChannelId.values(), //
+				StartStoppable.ChannelId.values(), //
+				BatteryInverterSmaStpSe.ChannelId.values() //
+		);
+
+	}
+
+	@Reference(policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MANDATORY)
+	protected void setModbus(BridgeModbus modbus) {
+		super.setModbus(modbus);
+	}
+
+	@Activate
+	void activate(ComponentContext context, Config config) throws OpenemsException {
+		this.config = config;
+		if (super.activate(context, config.id(), config.alias(), config.enabled(), config.modbusUnitId(), this.cm,
+				"Modbus", config.modbus_id(), READ_FROM_MODBUS_BLOCK)) {
+			return;
+		}
+		this._setGridMode(GridMode.ON_GRID);
+		this._setInitializing(true);
+	}
+
+	@Override
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	@Override
+	protected void onSunSpecInitializationCompleted() {
+		this.logInfo(this.log, "SunSpec initialization finished. " + this.channels().size() + " Channels available.");
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.ACTIVE_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S103.W);
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.REACTIVE_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S103.V_AR);
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.MAX_APPARENT_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S121.W_MAX);
+
+
+		this.getActivePowerChannel().onSetNextValue(val -> {		
+				this._setDcDischargePower(TypeUtils.subtract(val.get(), this.getDcPvPower()));
+		});
+
+		this._setInitializing(false);
+	}
+
+	@Override
+	public void run(Battery battery, int setActivePower, int setReactivePower) throws OpenemsNamedException {
+		this.calculateEnergy();
+		if (this.config.readOnly()) {
+			return;
+		}
+		if (battery instanceof SmaBattery sma) {
+			sma.applyPower(setActivePower, setReactivePower);
+			return;
+		}
+		this.logError(this.log, "Failed to run inverter. Battery is not SMA battery.");
```

Comment:
Please remove log output and replace with a new fault state channel, e.g. WRONG_BATTERY